test/unittest/rtps/common/TimeTests.cpp

```diff
@@ -0,0 +1,194 @@
+    ss.clear();
+
+    Time_t t3(4346, 0);
+    t3.nanosec(415672901);
+    ss << t3;
+    ASSERT_EQ(ss.str(), "4346.415672901");
+    ss.str("");
+    ss.clear();
+
+    // double constructor
+    Time_t t4(2.0);
+    ss << t4;
+    ASSERT_EQ(ss.str(), "2.0");
+    ss.str("");
+    ss.clear();
+
+    Time_t t5(0.000000002);
+    ss << t5;
+    ASSERT_EQ(ss.str(), "0.000000002");
+    ss.str("");
+    ss.clear();
+
+    // sec & frac constructor
+    Time_t t6(4346, 5);
+    ss << t6;
+    ASSERT_EQ(ss.str(), "4346.000000001");
+    ss.str("");
+    ss.clear();
+}
+
+/*
+ * Test >> operator
+ */
+TEST(TimeTest, deserialize_operator)
+{
+    Time_t t;
+
+    // This does not work due to double precision
+    // std::stringstream st1("1.02");
+    // st1 >> t;
+    // ASSERT_EQ(t.seconds(), 1);
+    // ASSERT_EQ(t.nanosec(), 20000000u);
+
+    std::stringstream st1("1.002");
+    st1 >> t;
+    ASSERT_EQ(t.seconds(), 1);
+    ASSERT_EQ(t.nanosec(), 2000000u);
+
+    std::stringstream st2("0.000003");
+    st2 >> t;
+    ASSERT_EQ(t.seconds(), 0);
+    ASSERT_EQ(t.nanosec(), 3000u);
+
+    // Non nanosecs
+    std::stringstream st3("12");
+    st3 >> t;
+    ASSERT_EQ(t.seconds(), 12);
+    ASSERT_EQ(t.nanosec(), 0u);
+
+    // Lowest time
+    std::stringstream st4("0.000000001");
+    st4 >> t;
+    ASSERT_EQ(t.seconds(), 0);
+    ASSERT_EQ(t.nanosec(), 1u);
+
+    // Check double precision
+    std::stringstream st5("4346.415672901");
+    st5 >> t;
+    ASSERT_EQ(t.seconds(), 4346);
+    ASSERT_EQ(t.nanosec(), 415672901u);
+
+    std::stringstream st6("123.456789");
+    st6 >> t;
+    ASSERT_EQ(t.seconds(), 123);
+    ASSERT_EQ(t.nanosec(), 456789000u);
+}
+
+/*
+ * Test >> operator with bad format
+ */
+TEST(TimeTest, bad_format_deserialize_operator)
+{
+    Time_t t;
+
+    // Check more than 9 digits
+    {
+        std::stringstream st("1.1231231231");
+        st >> t;
+        ASSERT_EQ(t.seconds(), 1);
+        ASSERT_EQ(t.nanosec(), 123123123u);
+    }
+
+    {
+        std::stringstream st("2.789789789789");
+        st >> t;
+        ASSERT_EQ(t.seconds(), 2);
+        ASSERT_EQ(t.nanosec(), 789789789u);
+    }
+
+    {
+        std::stringstream st("1234.999999999999");
+        st >> t;
+        ASSERT_EQ(t.seconds(), 1234);
+        ASSERT_EQ(t.nanosec(), 999999999u);
+    }
+
+    {
+        std::stringstream st("55.1000000001");
+        st >> t;
+        ASSERT_EQ(t.seconds(), 55);
+        ASSERT_EQ(t.nanosec(), 100000000u);
+    }
+
+    // Check if incorrect format, the time is not set
+    // Negative case
+    {
+        std::stringstream st("-3");
+        st >> t;
+        ASSERT_EQ(t.seconds(), 0);
+        ASSERT_EQ(t.nanosec(), 0u);
+    }
+
+    // Non numeric case
+    {
+        std::stringstream st("non_number");
+        st >> t;
+        ASSERT_EQ(t.seconds(), 0);
+        ASSERT_EQ(t.nanosec(), 0u);
+    }
+
+    // Non numeric case after number
+    {
+        std::stringstream st("2non_number");
```

Comment:
This fails in Mac due to the following: https://bugs.llvm.org/show_bug.cgi?id=17782

The >> operator is not reliable when extracting doubles from a stream when characters may be also present.

Reply:
Is this behaviour really the appropriate for this method? Shouldn't an incorrect format return a "non valid" (probably 0.0) time?